webhooks-extension/pkg/endpoints/webhook.go

```diff
@@ -17,32 +17,287 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	restful "github.com/emicklei/go-restful"
+	logging "github.com/tektoncd/experimental/webhooks-extension/pkg/logging"
+	pipelinesv1alpha1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	v1alpha1 "github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"net/http"
 	"os"
 	"strconv"
 	"strings"
 	"sync"
+	"time"
+)
 
-	logging "github.com/tektoncd/experimental/webhooks-extension/pkg/logging"
+var modifyingConfigMapLock sync.Mutex
 
-	restful "github.com/emicklei/go-restful"
-	eventapi "github.com/knative/eventing-sources/pkg/apis/sources/v1alpha1"
+const eventListenerName = "tekton-webhooks-eventlistener"
 
-	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-)
+/*
+	Creation of the eventlistener, called when no eventlistener exists at
+	the point of webhook creation.
+*/
+func (r Resource) createEventListener(webhook webhook, namespace, monitorTriggerName string) (*v1alpha1.EventListener, error) {
+	hookParams, monitorParams := r.getParams(webhook)
 
-var modifyingConfigMapLock sync.Mutex
+	pushTrigger := r.newTrigger(webhook.Name+"-"+webhook.Namespace+"-push-event",
+		webhook.Pipeline+"-push-binding",
+		webhook.Pipeline+"-template",
+		webhook.GitRepositoryURL,
+		"push",
+		webhook.AccessTokenRef,
+		hookParams)
+
+	pullRequestTrigger := r.newTrigger(webhook.Name+"-"+webhook.Namespace+"-pullrequest-event",
+		webhook.Pipeline+"-pullrequest-binding",
+		webhook.Pipeline+"-template",
+		webhook.GitRepositoryURL,
+		"pull_request",
+		webhook.AccessTokenRef,
+		hookParams)
+
+	monitorTrigger := r.newTrigger(monitorTriggerName,
+		webhook.PullTask+"-binding",
+		webhook.PullTask+"-template",
+		webhook.GitRepositoryURL,
+		"pull_request",
+		webhook.AccessTokenRef,
+		monitorParams)
+
+	triggers := []v1alpha1.EventListenerTrigger{pushTrigger, pullRequestTrigger, monitorTrigger}
+
+	eventListener := v1alpha1.EventListener{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      eventListenerName,
+			Namespace: namespace,
+		},
+		Spec: v1alpha1.EventListenerSpec{
+			ServiceAccountName: os.Getenv("SERVICE_ACCOUNT"),
+			Triggers:           triggers,
+		},
+	}
+
+	return r.TriggersClient.TektonV1alpha1().EventListeners(namespace).Create(&eventListener)
+}
+
+/*
+	Update of the eventlistener, called when adding additional webhooks as we
+	run with a single eventlistener.
+*/
+func (r Resource) updateEventListener(eventListener *v1alpha1.EventListener, webhook webhook, monitorTriggerName string) (*v1alpha1.EventListener, error) {
+	hookParams, monitorParams := r.getParams(webhook)
+	newPushTrigger := r.newTrigger(webhook.Name+"-"+webhook.Namespace+"-push-event",
```

Comment:
Bit of deja vu here or code duplication? Worth making the one utility function just so we don't have to maintain several blocks at once if we have to change something?

Reply:
im leaving as is ... if someone else wants to change it later they can have at it but I think it reads ok.